### PR TITLE
Change Collection::slice() method to return an instance of Collection

### LINKED ISSUE #195: slice() method should return a Collection instead of an array
See question here where this was originally requested https://github.com/doctrine/collections/pull/125#issuecomment-378990965

### PR DESCRIPTION
**Pull request by jwage**
Fixes #195

### CODE REVIEW

File: lib/Doctrine/Common/Collections/Collection.php [hunk 1]
```diff
@@ -286,9 +286,9 @@ public function indexOf($element);
      * @param int      $offset The offset to start from.
      * @param int|null $length The maximum number of elements to return, or null for no limit.
      *
-     * @return array
+     * @return Collection
      *
-     * @psalm-return array<TKey,T>
+     * @psalm-return Collection<TKey, T>
```
Comment (greg0ire) [inline lib/Doctrine/Common/Collections/Collection.php]: Should we add `@return static` here, and then change the type hints of implementors to `self` when dropping php 7.3?

Comment (jwage) [reply lib/Doctrine/Common/Collections/Collection.php]: I am not sure. Can that be done in another issue/PR?

File: lib/Doctrine/Common/Collections/AbstractLazyCollection.php [hunk 2]
```diff
@@ -268,7 +268,7 @@ public function indexOf($element)
     /**
      * {@inheritDoc}
      */
-    public function slice(int $offset, ?int $length = null) : array
+    public function slice(int $offset, ?int $length = null) : Collection
```
Comment (Ocramius) [inline lib/Doctrine/Common/Collections/AbstractLazyCollection.php]: What is the impact of this BC break on the OSS ecosystem? Is `->slice()` called often?

Comment (jwage) [reply lib/Doctrine/Common/Collections/AbstractLazyCollection.php]: I am not sure. Trying to think of a way to search for that on github somehow.

File: lib/Doctrine/Common/Collections/Collection.php (hunk 1 above)
Comment (greg0ire) [reply lib/Doctrine/Common/Collections/Collection.php]: Sure! It's a tiny detail, definitely not a blocker

**Comment by greg0ire** [comment]
Oh I just noticed this has been taken over already.

Closing in favor of https://github.com/doctrine/collections/pull/264

**Pull request closed without merging**